pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java

```diff
@@ -564,7 +538,11 @@ public boolean isIndexUseAllowed(DataSource dataSource, FieldConfig.IndexType in
   }
 
   public boolean isUnsafeTrim() {
-    return !isSameOrderAndGroupByColumns(this) || getHavingFilter() != null;
+    if (_isUnsafeTrim != null) {
```

Comment:
We usually pre-compute this during the `build()` call. We can do the same.
Currently this can still be computed multiple times